modules/engine-prototype/src/main/scala/engine/Step.scala

```diff
@@ -0,0 +1,251 @@
+package engine
+
+import scala.concurrent.ExecutionContext
+
+import cats.implicits._
+import cats.effect.Effect
+
+sealed trait Step
+object Step {
+
+  // Whatever is shared among all Steps. Notice it doesn't // extend Step
+  // because it's meant to be used only internally
+  case class Core(
+    id: String,
+    config: Core.Configuration,
+    breakpoint: Boolean
+  )
+
+  object Core {
+    case class Configuration()
+  }
+
+  sealed trait F2 extends Step
+  object F2 {
+
+    case class Done(core: Core, fileId: String) extends F2
+
+    sealed trait Pending extends F2
+    object Pending {
+      case class Standard(core: Core) extends Pending
+      case class Flat(core: Core) extends Pending
+      case class Dark(core: Core) extends Pending
+    }
+
+    sealed trait Ongoing extends F2
+    object Ongoing {
+      final case class Standard(core: Core, tcsConfigured: Boolean, instConfigured: Boolean) extends Ongoing
+    }
+
+    sealed trait Failed extends F2
+    object Failed {
+      final case class Standard(core: Core, tcsFailed: Boolean, instFailed: Boolean, observationFailed: Boolean) extends F2
+      // TODO: final case class Standard(core: Core, tcsFailure: Option[Failure], instFailure: Option[Failure]) extends F2
+    }
+
+    sealed trait Current extends F2 {
+
+      // Accessor to the core Step
+      val core: Core = ???
```

Comment:
Lots of accessors like this would have to be written in order to use such a deeply nested ADTs, so I'm still unsure if there is a better way to encode the ADTs.